### PR TITLE
Add WhenAll combinator

### PR DESCRIPTION
**Pull request by kononovk**
Add WhenAll for futures #27

### CODE REVIEW

File: include/yaclib/async/when_all.hpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    if (state == util::ResultState::Value) {
+      const auto ticket = _ticket.fetch_add(1, std::memory_order_acq_rel);
+      _results[ticket] = std::move(result).Value();
+      return;
+    }
+    if (_done.exchange(true, std::memory_order_acq_rel)) {
+      return;
+    }
+    if (state == util::ResultState::Error) {
+      std::move(_promise).Set(std::move(result).Exception());
```
Comment (Ri7ay) [inline include/yaclib/async/when_all.hpp:38]: Change to 
`std::move(_promise).Set(std::move(result).Error());`

File: include/yaclib/async/when_all.hpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+          std::move(promise).Set(std::vector<T>{});
+        }
+        return std::pair<Future<FutureValue>, container::intrusive::Ptr<AllCombinator>>{std::move(future), 0};
+      }
+      return std::pair<Future<FutureValue>, container::intrusive::Ptr<AllCombinator>>{
+          std::move(future), new container::Counter<AllCombinator>{std::move(promise), size}};
+    }
+  }
+
+  void Combine(util::Result<T> result) {
```
Comment (MBkkt) [inline include/yaclib/async/when_all.hpp:47]: - [ ] Add noexcept

File: include/yaclib/async/when_all.hpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+ private:
+  alignas(kCacheLineSize) std::atomic<bool> _done;
+  Promise<FutureValue> _promise;
+};
+
+template <size_t N, typename T, typename... Fs>
+void WhenAllImpl(container::intrusive::Ptr<AllCombinator<T, N>>& combinator, Future<T>&& head, Fs&&... tail) {
+  std::move(head).Subscribe([c = combinator](util::Result<T> result) mutable {
+    c->Combine(std::move(result));
+    c = nullptr;
```
Comment (MBkkt) [inline include/yaclib/async/when_all.hpp:94]: - [ ] remove this line

File: include/yaclib/async/when_all.hpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        if constexpr (std::is_void_v<T>) {
+          std::move(promise).Set();
+        } else {
+          std::move(promise).Set(std::vector<T>{});
+        }
+        return std::pair<Future<FutureValue>, container::intrusive::Ptr<AllCombinator>>{std::move(future), 0};
+      }
+      return std::pair<Future<FutureValue>, container::intrusive::Ptr<AllCombinator>>{
+          std::move(future), new container::Counter<AllCombinator>{std::move(promise), size}};
+    }
```
Comment (MBkkt) [inline include/yaclib/async/when_all.hpp:44]:
```suggestion
    if constexpr (!IsArray) {
      if (size == 0) {
        if constexpr (std::is_void_v<T>) {
          std::move(promise).Set();
        } else {
          std::move(promise).Set(std::vector<T>{});
        }
        return {std::move(future), nullptr};
      }
    }
    return {std::move(future), new container::Counter<AllCombinator>{std::move(promise), size}};

```

**Review by MBkkt — APPROVED** [review]

**Pull request merged**